src/libraries/Common/tests/System/Net/Prerequisites/NetCoreServer/Handlers/EchoWebSocketHandler.cs

```diff
@@ -144,6 +144,27 @@ public static async Task InvokeAsync(HttpContext context)
                     {
                         await Task.Delay(5000);
                     }
+                    else if (receivedMessage == ".shutdownAfterTwoMessages")
+                    {
+                        byte[] buffer = new byte[1024];
+                        string message = DateTime.Now.ToString("HH:mm:ss");
+                        buffer = System.Text.Encoding.UTF8.GetBytes(message);
+                        await socket.SendAsync(
+                            new ArraySegment<byte>(buffer, 0, message.Length),
+                            WebSocketMessageType.Text,
+                            true,
+                            CancellationToken.None);
+                        await Task.Delay(1990); // try to sync with receive request from the client: 1.9k is too little, 2k too much
```

Comment:
This is the tricky part. What would happen if the wait was not there at all ?

Reply:
We would have sent one after another, so it would be event order:
- connect, ask for receive 1, send1, send2, ask for receive 2.

Adding this delay we are on the boarder of the above scenario and the below:
- connect, ask for receive 1, send1, ask for receive 2, send 2.

By "on the boarder" I mean it's random, sometimes we fall into 1st scenario, sometimes into the 2nd.

Reply:
`ask for receive 2` should have no impact on the JS side order of WS events and JS side buffering, right ?

Reply:
Are we trying to get into situation that "ws.state == CLOSE" but the `on_message` for the second message was not called yet ?

If that's possible, the current implementation is wrong. But I need to see it before I can believe it 🔍

Reply:
The time when we ask for receive seems important, when using the debugger, we enter the receiving method before we got "on_message_sent" event but after the WS is closed. That's why I was trying to time it as close in time to closing and sending actions as possible.

Reply:
Yes, closing before sending seems impossible but that's the order of events that JS records.

Reply:
I think that debugger is "lying", because it could see the `ws.state` after browser internals received the close, but because you have breakpoint it was not able to deliver it yet. ??

Reply:
Anyway, we could try to stop using ws.state for this and rely on the `on_close` event

Reply:
Removing the `if (readyState == WebSocket.CLOSED)` block of code from `ws_wasm_receive` eliminates the exception.

Reply:
Does it mean that
```
     const readyState = ws.readyState;
     if (readyState == WebSocket.CLOSED) {
         mono_log_warn("ws_wasm_receive unsafe block of code reached, readyState == WebSocket.CLOSED");
         const receive_status_ptr = ws[wasm_ws_receive_status_ptr];
         setI32(receive_status_ptr, 0); // count
         setI32(<any>receive_status_ptr + 4, 2); // type:close
         setI32(<any>receive_status_ptr + 8, 1);// end_of_message: true
         return resolvedPromise();
     }

```
is redundant? Managed code does no need setting these bits?

Reply:
That's good question. `on_close` has similar code and would deliver the resolution if we deleted this.

Reply:
There is a difference between the demo and the tests: demo does not close the socket, only dispose of it and in case the socket was opened without using `CloseAsync()`, the WebSocket connection will be abruptly terminated. This termination will occur without following the WebSocket protocol's normal closure handshake, potentially leading to unexpected behavior or errors on the client side. Upon correcting the demo code to close the socket, not only dispose of it, I cannot reproduce the issue with "lost messages". The current, working version of code has the `if (readyState == WebSocket.CLOSED)` block removed.